### PR TITLE
chore(weave): refactor cost query

### PR DESCRIPTION
**Pull request by jwlee64**
refactors the cost query from something like
  orig cost query
  get usage from cost - 1 req to calls_merged
  get ranked prices
  discard not ranked prices
  join with usage to get cost - 1 req to calls_merged
  final select join with calls and construct cost - 1 req to calls_merged
  
to
  orig cost query
  get usage from cost - 1 req to calls_merged
  get ranked prices
  final select group by construct cost 
  
takes about 93k->31k rows scanned
unsure about latency speed up exactly

### CODE REVIEW

File: weave/trace_server/token_costs.py [hunk 1]
```diff
@@ -18,8 +18,6 @@
 cost_numeric_fields = [
     "prompt_token_cost",
     "completion_token_cost",
-    "prompt_tokens_cost",
```
Comment (tssweeney) [inline weave/trace_server/token_costs.py:21]: Does this interface need modifications??

```
class LLMCostSchema(LLMUsageSchema):
    prompt_tokens_cost: Optional[float]
    completion_tokens_cost: Optional[float]
    prompt_token_cost: Optional[float]
    completion_token_cost: Optional[float]
 ```

File: weave/trace_server/token_costs.py [hunk 2]
```diff
@@ -399,40 +258,34 @@ def final_call_select_with_cost(
     """
 
     summary_dump_snippet = f"""
+    if( any(llm_id) = 'nothing',
```
Comment (tssweeney) [inline weave/trace_server/token_costs.py:261]: where does 'nothing' come from??

File: weave/trace_server/tests/test_calls_query_builder.py [hunk 3]
```diff
@@ -292,124 +292,123 @@ def test_query_light_column_with_costs() -> None:
...
+                                ELSE 2
+                            END,
+                            CASE
+                                -- Order by pricing level then by effective_date
                                 -- WHEN llm_token_prices.pricing_level = 'org' AND llm_token_prices.pricing_level_id = ORG_NAME THEN 1
-                                WHEN llm_token_prices.pricing_level = 'project'
-                                    AND llm_token_prices.pricing_level_id = 'project' THEN 2
-                                WHEN llm_token_prices.pricing_level = 'default'
-                                    AND llm_token_prices.pricing_level_id = 'default' THEN 3
+                                WHEN llm_token_prices.pricing_level = 'project' AND llm_token_prices.pricing_level_id = 'project' THEN 2
```
Comment (tssweeney) [inline weave/trace_server/tests/test_calls_query_builder.py:356]: `llm_token_prices.pricing_level_id = 'project'` be parameterized?

**Review by tssweeney — APPROVED** [review]

File: weave/trace_server/token_costs.py [hunk 4]
```diff
@@ -441,39 +372,69 @@ def final_call_select_with_cost(
...
+# This checks that a project_id is a valid base64 encoded string, that follows the pattern "ProjectInternalId: <number>"
+def validate_project_id(project_id: str) -> None:
+    try:
+        # Attempt to decode the id from Base64
+        decoded_str = base64.b64decode(project_id).decode("utf-8")
+
+        # Check if the decoded id matches the pattern "ProjectInternalId:" followed by a number
+        match = (
+            re.fullmatch(r"ProjectInternalId:\d+", decoded_str.strip())
+            or decoded_str == "shawn/test-project"
```
Comment (jwlee64) [inline weave/trace_server/token_costs.py:432]: this is the entity we use in our tests

File: weave/trace_server/token_costs.py [hunk 5]
```diff
@@ -187,20 +188,28 @@ def get_ranked_prices(
...
 
-    # Select fields
-    lu_fields = [f"{llm_usage_table_alias}.{col.name}" for col in llm_usage_cols]
     ltp_fields = [
         f"{LLM_TOKEN_PRICES_TABLE.name}.{col.name}" for col in LLM_TOKEN_PRICES_COLUMNS
     ]
+
+    # Clickhouse does not allow parameters in the row_number() over function
+    # This is a temporary workaround, to check the validity of the project_id, to prevent SQL injection
+    validate_project_id(project_id)
```
Comment (jwlee64) [inline weave/trace_server/token_costs.py:197]: where we validate the project_id

Comment (tssweeney) [reply weave/trace_server/token_costs.py:197]: I would be even more explicit here: `sql_injection_safe_project_id = ...`

**Review by tssweeney — APPROVED** [review]

**Pull request merged**